## ETL del dataset australian_user_reviews

A continuación se realiza la extracción, transformación y carga del conjunto de datos australian_user_reviews

### Importaciones

In [2]:
import pandas as pd
import ast

%load_ext autoreload
%autoreload 2
import funciones

import warnings
warnings.filterwarnings("ignore")

Extracción, primera exploración y eliminación de duplicados

In [3]:
# Ruta al dataset australian_user_reviews
ruta_review = 'C:/Users/Gabriela/Documents/GitHub/PI_ML_OPS/data/australian_user_reviews.json'


# Se lee de cada línea del dataset
filas_review = []
with open(ruta_review) as f:
    for line in f.readlines():
        filas_review.append(ast.literal_eval(line))

# Se convierte en dataframe
df_reviews = pd.DataFrame(filas_review)
df_reviews

user_id                                           user_url  \
0      76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1                js41637               http://steamcommunity.com/id/js41637   
2              evcentric             http://steamcommunity.com/id/evcentric   
3                  doctr                 http://steamcommunity.com/id/doctr   
4              maplemage             http://steamcommunity.com/id/maplemage   
...                  ...                                                ...   
25794  76561198306599751  http://steamcommunity.com/profiles/76561198306...   
25795           Ghoustik              http://steamcommunity.com/id/Ghoustik   
25796  76561198310819422  http://steamcommunity.com/profiles/76561198310...   
25797  76561198312638244  http://steamcommunity.com/profiles/76561198312...   
25798        LydiaMorley           http://steamcommunity.com/id/LydiaMorley   

                                                 reviews  
0      [{'funny': '', 'posted': 'Posted November 5, 2...  
1      [{'funny': '', 'posted': 'Posted June 24, 2014...  
2      [{'funny': '', 'posted': 'Posted February 3.',...  
3      [{'funny': '', 'posted': 'Posted October 14, 2...  
4      [{'funny': '3 people found this review funny',...  
...                                                  ...  
25794  [{'funny': '', 'posted': 'Posted May 31.', 'la...  
25795  [{'funny': '', 'posted': 'Posted June 17.', 'l...  
25796  [{'funny': '1 person found this review funny',...  
25797  [{'funny': '', 'posted': 'Posted July 21.', 'l...  
25798  [{'funny': '1 person found this review funny',...  

[25799 rows x 3 columns]

In [4]:
# Obtener información general del DataFrame
df_reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25799 entries, 0 to 25798
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   user_id   25799 non-null  object
 1   user_url  25799 non-null  object
 2   reviews   25799 non-null  object
dtypes: object(3)
memory usage: 604.8+ KB


In [5]:
#Se verifican los tipos de datos para cada columna
funciones.tipo_de_datos(df_reviews)

nombre_campo        tipo_datos  no_nulos_%  nulos_%  nulos
0      user_id   [<class 'str'>]       100.0      0.0      0
1     user_url   [<class 'str'>]       100.0      0.0      0
2      reviews  [<class 'list'>]       100.0      0.0      0

De la anterior revisión se extrae como información la no presencia de datos nulos en el dataset. Se procede a chequear la presencia de datos duplicados.

In [6]:
# Se chequean datos duplicados en la columna 'user_id'
funciones.duplicados(df_reviews, 'user_id')

user_id  \
12888               05041129   
5250                05041129   
3133   111222333444555666888   
3134   111222333444555666888   
4139                   29123   
...                      ...   
2721          xXAussieRockXx   
2680             yolofaceguy   
17916            yolofaceguy   
5855               zeroblade   
13975              zeroblade   

                                                user_url  \
12888              http://steamcommunity.com/id/05041129   
5250               http://steamcommunity.com/id/05041129   
3133   http://steamcommunity.com/id/11122233344455566...   
3134   http://steamcommunity.com/id/11122233344455566...   
4139                  http://steamcommunity.com/id/29123   
...                                                  ...   
2721         http://steamcommunity.com/id/xXAussieRockXx   
2680            http://steamcommunity.com/id/yolofaceguy   
17916           http://steamcommunity.com/id/yolofaceguy   
5855              http://steamcommunity.com/id/zeroblade   
13975             http://steamcommunity.com/id/zeroblade   

                                                 reviews  
12888  [{'funny': '', 'posted': 'Posted May 18, 2015....  
5250   [{'funny': '', 'posted': 'Posted May 18, 2015....  
3133   [{'funny': '', 'posted': 'Posted December 22, ...  
3134   [{'funny': '', 'posted': 'Posted December 22, ...  
4139   [{'funny': '', 'posted': 'Posted March 26.', '...  
...                                                  ...  
2721   [{'funny': '', 'posted': 'Posted July 17, 2015...  
2680   [{'funny': '', 'posted': 'Posted October 31, 2...  
17916  [{'funny': '', 'posted': 'Posted October 31, 2...  
5855   [{'funny': '', 'posted': 'Posted November 30, ...  
13975  [{'funny': '', 'posted': 'Posted November 30, ...  

[623 rows x 3 columns]

Se observan 623 filas duplicadas. Por lo que se decide eliminarlas.

In [7]:
df_reviews = df_reviews.drop_duplicates(subset='user_id', keep='first')

#Se chequea si han quedado registros duplicados.
funciones.duplicados(df_reviews, 'user_id')

'No hay duplicados'

A continuación se decide revisar la columna 'reviews'

In [8]:
df_reviews['reviews'][0]

[{'funny': '',
  'posted': 'Posted November 5, 2011.',
  'last_edited': '',
  'item_id': '1250',
  'helpful': 'No ratings yet',
  'recommend': True,
  'review': 'Simple yet with great replayability. In my opinion does "zombie" hordes and team work better than left 4 dead plus has a global leveling system. Alot of down to earth "zombie" splattering fun for the whole family. Amazed this sort of FPS is so rare.'},
 {'funny': '',
  'posted': 'Posted July 15, 2011.',
  'last_edited': '',
  'item_id': '22200',
  'helpful': 'No ratings yet',
  'recommend': True,
  'review': "It's unique and worth a playthrough."},
 {'funny': '',
  'posted': 'Posted April 21, 2011.',
  'last_edited': '',
  'item_id': '43110',
  'helpful': 'No ratings yet',
  'recommend': True,
  'review': 'Great atmosphere. The gunplay can be a bit chunky at times but at the end of the day this game is definitely worth it and I hope they do a sequel...so buy the game so I get a sequel!'}]

### Transformación

#### Transformación de la columna 'reviews'

De lo anterior, se concluye que la columna 'reviews' se presenta anidada, siendo una lista con uno o mas diccionarios como elementos.Por lo tanto, se busca generar una columna por cada diccionario para posteriormente hacer un registro por cada diccionario.

In [9]:
# Se transforma a columnas cada elemento de las listas
df_reviews2 = pd.json_normalize(df_reviews['reviews'])
df_reviews2.head()

0  \
0  {'funny': '', 'posted': 'Posted November 5, 20...   
1  {'funny': '', 'posted': 'Posted June 24, 2014....   
2  {'funny': '', 'posted': 'Posted February 3.', ...   
3  {'funny': '', 'posted': 'Posted October 14, 20...   
4  {'funny': '3 people found this review funny', ...   

                                                   1  \
0  {'funny': '', 'posted': 'Posted July 15, 2011....   
1  {'funny': '', 'posted': 'Posted September 8, 2...   
2  {'funny': '', 'posted': 'Posted December 4, 20...   
3  {'funny': '', 'posted': 'Posted July 28, 2012....   
4  {'funny': '1 person found this review funny', ...   

                                                   2  \
0  {'funny': '', 'posted': 'Posted April 21, 2011...   
1  {'funny': '', 'posted': 'Posted November 29, 2...   
2  {'funny': '', 'posted': 'Posted November 3, 20...   
3  {'funny': '', 'posted': 'Posted June 2, 2012.'...   
4  {'funny': '2 people found this review funny', ...   

                                                   3  \
0                                               None   
1                                               None   
2  {'funny': '', 'posted': 'Posted October 15, 20...   
3  {'funny': '', 'posted': 'Posted June 29, 2014....   
4  {'funny': '', 'posted': 'Posted July 11, 2013....   

                                                   4  \
0                                               None   
1                                               None   
2  {'funny': '', 'posted': 'Posted October 15, 20...   
3  {'funny': '', 'posted': 'Posted November 22, 2...   
4                                               None   

                                                   5     6     7     8     9  
0                                               None  None  None  None  None  
1                                               None  None  None  None  None  
2  {'funny': '', 'posted': 'Posted October 15, 20...  None  None  None  None  
3  {'funny': '', 'posted': 'Posted February 23, 2...  None  None  None  None  
4                                               None  None  None  None  None

Con lo anterior, se pierde la columna 'user_id' y 'user_url' pero se mantienen las posiciones, por lo cual, a continuación, se concatenarán ambos dataframes para poder tener nuevamente disponible la trazabilidad de usuario.

In [10]:
df_reviews2 = pd.concat([df_reviews[['user_id', 'user_url']], df_reviews2], axis=1)
df_reviews2.head()

user_id                                           user_url  \
0  76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1            js41637               http://steamcommunity.com/id/js41637   
2          evcentric             http://steamcommunity.com/id/evcentric   
3              doctr                 http://steamcommunity.com/id/doctr   
4          maplemage             http://steamcommunity.com/id/maplemage   

                                                   0  \
0  {'funny': '', 'posted': 'Posted November 5, 20...   
1  {'funny': '', 'posted': 'Posted June 24, 2014....   
2  {'funny': '', 'posted': 'Posted February 3.', ...   
3  {'funny': '', 'posted': 'Posted October 14, 20...   
4  {'funny': '3 people found this review funny', ...   

                                                   1  \
0  {'funny': '', 'posted': 'Posted July 15, 2011....   
1  {'funny': '', 'posted': 'Posted September 8, 2...   
2  {'funny': '', 'posted': 'Posted December 4, 20...   
3  {'funny': '', 'posted': 'Posted July 28, 2012....   
4  {'funny': '1 person found this review funny', ...   

                                                   2  \
0  {'funny': '', 'posted': 'Posted April 21, 2011...   
1  {'funny': '', 'posted': 'Posted November 29, 2...   
2  {'funny': '', 'posted': 'Posted November 3, 20...   
3  {'funny': '', 'posted': 'Posted June 2, 2012.'...   
4  {'funny': '2 people found this review funny', ...   

                                                   3  \
0                                               None   
1                                               None   
2  {'funny': '', 'posted': 'Posted October 15, 20...   
3  {'funny': '', 'posted': 'Posted June 29, 2014....   
4  {'funny': '', 'posted': 'Posted July 11, 2013....   

                                                   4  \
0                                               None   
1                                               None   
2  {'funny': '', 'posted': 'Posted October 15, 20...   
3  {'funny': '', 'posted': 'Posted November 22, 2...   
4                                               None   

                                                   5     6     7     8     9  
0                                               None  None  None  None  None  
1                                               None  None  None  None  None  
2  {'funny': '', 'posted': 'Posted October 15, 20...  None  None  None  None  
3  {'funny': '', 'posted': 'Posted February 23, 2...  None  None  None  None  
4                                               None  None  None  None  None

Ahora que se tienen los diccionarios por columnas, con el usuario que genera dicha información, se genera un registro por cada diccionario, manteniendo en cada caso el usuario que lo genera.

In [11]:
# Se utiliza pd.melt para transformar las columnas en filas conservando el 'user_id' y 'user_url'
df_reviews2 = pd.melt(df_reviews2, id_vars=['user_id', 'user_url'], 
                       value_vars=list(range(9)),
                       value_name='reviews')
df_reviews2.head()

user_id                                           user_url  \
0  76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1            js41637               http://steamcommunity.com/id/js41637   
2          evcentric             http://steamcommunity.com/id/evcentric   
3              doctr                 http://steamcommunity.com/id/doctr   
4          maplemage             http://steamcommunity.com/id/maplemage   

  variable                                            reviews  
0        0  {'funny': '', 'posted': 'Posted November 5, 20...  
1        0  {'funny': '', 'posted': 'Posted June 24, 2014....  
2        0  {'funny': '', 'posted': 'Posted February 3.', ...  
3        0  {'funny': '', 'posted': 'Posted October 14, 20...  
4        0  {'funny': '3 people found this review funny', ...

Al realizar este proceso se puede ver que hay registros None, lo cual ocurre ya que hay usuarios que han realizado más reviews que otros.
Por lo tanto, se decide eliminar aquellas filas que contengan None.

In [12]:
# Se eliminan las filas con valor None
df_reviews2 = df_reviews2.dropna()

In [13]:
# Se separan por columnas cada una de las claves de 'reviews'
df_reviews = df_reviews2['reviews'].apply(pd.Series, dtype='object')
df_reviews = df_reviews.add_prefix('reviews_')
df_reviews.head()

reviews_funny            reviews_posted  \
0                                    Posted November 5, 2011.   
1                                       Posted June 24, 2014.   
2                                          Posted February 3.   
3                                    Posted October 14, 2013.   
4  3 people found this review funny    Posted April 15, 2014.   

  reviews_last_edited reviews_item_id  \
0                                1250   
1                              251610   
2                              248820   
3                              250320   
4                              211420   

                                   reviews_helpful  reviews_recommend  \
0                                   No ratings yet               True   
1  15 of 20 people (75%) found this review helpful               True   
2                                   No ratings yet               True   
3   2 of 2 people (100%) found this review helpful               True   
4  35 of 43 people (81%) found this review helpful               True   

                                      reviews_review  
0  Simple yet with great replayability. In my opi...  
1  I know what you think when you see this title ...  
2  A suitably punishing roguelike platformer.  Wi...  
3  This game... is so fun. The fight sequences ha...  
4                                            Git gud

En el procesamiento anterior, se puede ver que la columna de 'user_id' y 'user_url' se perdió nuevamente, por lo que se vuelve a concatenar.

In [14]:
# Se concatena con 'user_id' y 'user_url'
df_reviews = pd.concat([df_reviews2[['user_id', 'user_url']], df_reviews], axis=1)
df_reviews.head()

user_id                                           user_url  \
0  76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1            js41637               http://steamcommunity.com/id/js41637   
2          evcentric             http://steamcommunity.com/id/evcentric   
3              doctr                 http://steamcommunity.com/id/doctr   
4          maplemage             http://steamcommunity.com/id/maplemage   

                      reviews_funny            reviews_posted  \
0                                    Posted November 5, 2011.   
1                                       Posted June 24, 2014.   
2                                          Posted February 3.   
3                                    Posted October 14, 2013.   
4  3 people found this review funny    Posted April 15, 2014.   

  reviews_last_edited reviews_item_id  \
0                                1250   
1                              251610   
2                              248820   
3                              250320   
4                              211420   

                                   reviews_helpful  reviews_recommend  \
0                                   No ratings yet               True   
1  15 of 20 people (75%) found this review helpful               True   
2                                   No ratings yet               True   
3   2 of 2 people (100%) found this review helpful               True   
4  35 of 43 people (81%) found this review helpful               True   

                                      reviews_review  
0  Simple yet with great replayability. In my opi...  
1  I know what you think when you see this title ...  
2  A suitably punishing roguelike platformer.  Wi...  
3  This game... is so fun. The fight sequences ha...  
4                                            Git gud

En algunas columnas se encuentran valores faltantes, los cuales no figuran como None. Se procede a chequear si esto se debe a la presencia de espacios.

In [15]:
df_reviews['reviews_last_edited'][0]

''

In [16]:
# Se reemplzan estos valores por nulos.
df_reviews.replace('', None, inplace=True)

Se realiza un chequeo general sobre nuestro dataset.

In [17]:
funciones.tipo_de_datos(df_reviews)

nombre_campo                           tipo_datos  no_nulos_%  \
0              user_id                      [<class 'str'>]      100.00   
1             user_url                      [<class 'str'>]      100.00   
2        reviews_funny  [<class 'NoneType'>, <class 'str'>]       13.76   
3       reviews_posted                      [<class 'str'>]      100.00   
4  reviews_last_edited  [<class 'NoneType'>, <class 'str'>]       10.28   
5      reviews_item_id                      [<class 'str'>]      100.00   
6      reviews_helpful                      [<class 'str'>]      100.00   
7    reviews_recommend                     [<class 'bool'>]      100.00   
8       reviews_review  [<class 'str'>, <class 'NoneType'>]       99.95   

   nulos_%  nulos  
0     0.00      0  
1     0.00      0  
2    86.24  49498  
3     0.00      0  
4    89.72  51499  
5     0.00      0  
6     0.00      0  
7     0.00      0  
8     0.05     30

Se observa que las columnas 'reviews_funny' y 'review_last_edited' hay entre un 86% a 89% de datos faltantes. Se elimina así estas columnas.

In [18]:
# Se eliminan las columnas 'reviews_funny' y 'reviews_last_edited'
df_reviews = df_reviews.drop(columns=['reviews_funny', 'reviews_last_edited'])

####  Transformación de la columna 'reviews_posted'

Se procede a transformar el formato de la fecha en que se hizo el posteo de la review ya que se encuentra en formato 'str'. 

In [19]:
df_reviews['reviews_date'] = df_reviews['reviews_posted'].apply(funciones.formato_fecha)
df_reviews['reviews_date']

0               2011-11-05
1               2014-06-24
2         Formato inválido
3               2013-10-14
4               2014-04-15
                ...       
231291          2014-08-15
231293          2014-08-02
231419          2015-07-31
231499          2015-12-20
231501    Formato inválido
Name: reviews_date, Length: 57397, dtype: object

Se revisan aquellas reviews con formato inválido.

In [20]:
df_reviews[df_reviews['reviews_date'] == 'Formato inválido']

user_id                                           user_url  \
2               evcentric             http://steamcommunity.com/id/evcentric   
6       76561198079601835  http://steamcommunity.com/profiles/76561198079...   
7             MeaTCompany           http://steamcommunity.com/id/MeaTCompany   
9       76561198156664158  http://steamcommunity.com/profiles/76561198156...   
10      76561198077246154  http://steamcommunity.com/profiles/76561198077...   
...                   ...                                                ...   
223569  76561198040184950  http://steamcommunity.com/profiles/76561198040...   
226105  76561198046474248  http://steamcommunity.com/profiles/76561198046...   
228109         dmitry_who            http://steamcommunity.com/id/dmitry_who   
229231  76561198079507136  http://steamcommunity.com/profiles/76561198079...   
231501       EchoXSilence          http://steamcommunity.com/id/EchoXSilence   

            reviews_posted reviews_item_id  \
2       Posted February 3.          248820   
6           Posted May 20.             730   
7          Posted July 24.             730   
9          Posted June 16.          252950   
10         Posted June 11.             440   
...                    ...             ...   
223569    Posted April 12.          394690   
226105    Posted March 28.          234140   
228109      Posted May 17.          376210   
229231   Posted January 3.             730   
231501  Posted February 3.             570   

                                        reviews_helpful  reviews_recommend  \
2                                        No ratings yet               True   
6          0 of 1 people (0%) found this review helpful               True   
7                                        No ratings yet               True   
9          0 of 1 people (0%) found this review helpful               True   
10                                       No ratings yet               True   
...                                                 ...                ...   
223569                                   No ratings yet               True   
226105                                   No ratings yet               True   
228109  10 of 28 people (36%) found this review helpful               True   
229231                                   No ratings yet              False   
231501     0 of 2 people (0%) found this review helpful              False   

                                           reviews_review      reviews_date  
2       A suitably punishing roguelike platformer.  Wi...  Formato inválido  
6                                           ZIKA DO BAILE  Formato inválido  
7                           BEST GAME IN THE BLOODY WORLD  Formato inválido  
9                                                 love it  Formato inválido  
10                                                 mt bom  Formato inválido  
...                                                   ...               ...  
223569  I cannot say much right now due to the game no...  Formato inválido  
226105  Oh what a day .., What a lovely day to play th...  Formato inválido  
228109  ░░░░░░░░░░░█▀▀░░█░░░░░░░░░░░▄▀▀▀▀░░░░░█▄▄░░░░░...  Formato inválido  
229231                                          got VACed  Formato inválido  
231501  DOTA 2.normal skill - 2ez for me>high skill - ...  Formato inválido  

[9771 rows x 8 columns]

Una vez revisados los datos de la columna, se decide eliminar el la columna 'reviews_posted' ya que se considera que no aporta información relevante.

In [21]:
df_reviews = df_reviews.drop('reviews_posted', axis=1)

#### Transformación columna 'reviews_review'

Comenzamos revisando el contenido y tipo de datos de la columna

In [22]:
funciones.tipo_de_datos(df_reviews)

nombre_campo                           tipo_datos  no_nulos_%  \
0            user_id                      [<class 'str'>]      100.00   
1           user_url                      [<class 'str'>]      100.00   
2    reviews_item_id                      [<class 'str'>]      100.00   
3    reviews_helpful                      [<class 'str'>]      100.00   
4  reviews_recommend                     [<class 'bool'>]      100.00   
5     reviews_review  [<class 'str'>, <class 'NoneType'>]       99.95   
6       reviews_date                      [<class 'str'>]      100.00   

   nulos_%  nulos  
0     0.00      0  
1     0.00      0  
2     0.00      0  
3     0.00      0  
4     0.00      0  
5     0.05     30  
6     0.00      0

La columna 'reviews_review' presenta un 5% de datos nulos, por lo que se decide eliminarlos.

In [23]:
df_reviews= df_reviews.dropna(subset=['reviews_review'])

### Carga del dataset australian_user_review

Se guarda el archivo con el dataset limpio en formato .csv

In [24]:
archivo_limpio = 'user_review_limpio.csv'
df_reviews.to_csv(archivo_limpio, index=False, encoding='utf-8')
print(f'Se guardó el archivo {archivo_limpio}')

Se guardó el archivo user_review_limpio.csv
